### PR TITLE
HPCC-22662 Report a WARNING for an alien type better expressed as DATA

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
see ecl/regress (alien5/dataset16) for examples
@richardkchapman is this a reasonable change to target 7.2.x?
@shamser please review

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -2643,6 +2643,40 @@ void HqlGram::checkDefaultValueVirtualAttr(const attribute &errpos, IHqlExpressi
...
+    {
+        if (size->queryChild(0)->getOperator() == no_selfref)
+        {
+            IHqlExpression & topRecord = activeRecords.tos();
+            if (topRecord.numChildren())
+            {
+                IHqlExpression * lastField = topRecord.queryChild(topRecord.numChildren()-1);
+                if (lastField == size->queryChild(1))
+                {
+                    if (lastField->queryType()->getSize() == 4)
```
Comment (richardkchapman) [inline ecl/hql/hqlgram2.cpp]: Feels like it should be checking that it's a little-endian unsigned integer as well as that it's 4-bytes? Possibly could also accept signed...

**Review by shamser — COMMENTED** [review]
@ghalliday I haven't been able to locate documentation on the format of data types. There is a mention of supporting variable length DATA but there doesn't seem to specific data on the format.(I've check Programmer's Guide and Programmers reference.)  Do we need to update our documentation?

File: ecl/hql/hqlgram2.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlgram2.cpp]: It depends if you are looking for binary compatibility, or equivalence.  The row translation will make life interesting even if binary compatible.

**Comment by ghalliday** [comment]
Yes, we should probably document the format of each of the field types when it is stored in a row.

**Review by shamser — APPROVED** [review]
@ghalliday Reviewed Ok.

**Comment by ghalliday** [comment]
@richardkchapman to discuss when you're back

**Pull request merged**